R/dryRun.R

```diff
@@ -0,0 +1,175 @@
+#'   it can't help with: using an absolute path and using the wrong case.
+#'   Both of these will work locally but fail on the server. [lint()]
+#'   uses an alternative technique (static analysis) to detect many of these
+#'   cases.
+#'
+#' @inheritParams deployApp
+#' @param contentCategory Set this to `"site"` if you'd deploy with
+#'   [deploySite()]; otherwise leave as is.
+#' @export
+dryRun <- function(appDir = getwd(),
+                   envVars = NULL,
+                   appFiles = NULL,
+                   appFileManifest = NULL,
+                   appPrimaryDoc = NULL,
+                   contentCategory = NULL,
+                   quarto = NA) {
+  check_installed("callr")
+
+  appFiles <- listDeploymentFiles(
+    appDir,
+    appFiles = appFiles,
+    appFileManifest = appFileManifest
+  )
+
+  appMetadata <- appMetadata(
+    appDir = appDir,
+    appFiles = appFiles,
+    appPrimaryDoc = appPrimaryDoc,
+    quarto = quarto,
+    contentCategory = contentCategory,
+  )
+
+  # copy files to bundle dir to stage
+  cli::cli_alert_info("Bundling app")
+  bundleDir <- bundleAppDir(
+    appDir = appDir,
+    appFiles = appFiles,
+    appPrimaryDoc = appMetadata$appPrimaryDoc
+  )
+  defer(unlink(bundleDir, recursive = TRUE))
+
+  cli::cli_alert_info("Creating project specific library")
+  callr::r(
+    function() {
+      options(renv.verbose = FALSE)
+      renv::init()
+      renv::restore()
+    },
+    wd = bundleDir
+  )
+
+  # Add tracing code -------------------------------------------------
+  cli::cli_alert_info("Adding shims")
+
+  file.copy(
+    system.file("dryRunTrace.R", package = "rsconnect"),
+    file.path(bundleDir, "__rsconnect-dryRunTrace.R")
+  )
+  appendLines(
+    file.path(bundleDir, ".Rprofile"),
+    c("", 'source("__rsconnect-dryRunTrace.R")')
+  )
+
+  # Run ---------------------------------------------------------------
+  cli::cli_alert_info("Starting {appMetadata$appMode}")
+  if (appMetadata$appMode %in% c("rmd-shiny", "quarto-shiny", "shiny", "api")) {
+    cli::cli_alert_warning("Terminate the app to complete the dry run")
+  }
+
+  envVarNames <- setdiff(userEnvVars(), c(envVars, "PATH"))
+  envVarReset <- c(rep_named(envVarNames, ""), callr::rcmd_safe_env())
+
+  callr::r(
+    appRunner(appMetadata$appMode),
+    args = list(primaryDoc = appMetadata$appPrimaryDoc),
+    env = envVarReset,
+    wd = bundleDir,
+    show = TRUE
+  )
+  invisible()
+}
+
+appRunner <- function(appMode) {
+  switch(appMode,
+    "rmd-static" = ,
+    "rmd-shiny" = function(primaryDoc) {
+      rmarkdown::render(primaryDoc, quiet = TRUE)
+    },
+    "quarto-static" = ,
+    "quarto-shiny" = function(primaryDoc) {
+      quarto::quarto_render(primaryDoc, quiet = TRUE)
+    },
+    "shiny" = function(primaryDoc) {
+      shiny::runApp()
+    },
+    "api" = function(primaryDoc) {
+      plumber::pr_run(plumber::pr("plumber.R"))
+    },
+    cli::cli_abort("Content type {appMode} not currently supported")
```

Comment:
Rendered content and "static" -- we could start an HTML server against the `bundleDir` so they could see exactly what files they are about to deploy. Maybe that would help cases where folks forget to include some images/CSS.

The result of this would be that _every_ content type starts some type of HTTP server that they would interact with for validation.